File: leetcode/1-100/46-permutations/bench/permutations.py
# LeetCode #46 bench mirror — Python, the used-array mutable-path backtracker (star).
#
# Mirrors bench/permutations.kara: a DFS that picks any still-unused element (tracked by a
# `used` bool list) alongside a mutable path, snapshotting into a list-of-lists at each leaf.
# Same workload (TOTAL permutations of a fixed n=7 array, one slot punched per iteration) and
# the same position-weighted checksum as every other mirror. Slow lane — the interpreted
# reference time.

TOTAL = 300
MODULUS = 1000000007
N = 7


def backtrack(nums, used, path, out):
    n = len(nums)
    if len(path) == n:
        out.append(path.copy())
        return
    for i in range(n):
        if not used[i]:
            used[i] = True
            path.append(nums[i])
            backtrack(nums, used, path, out)
            path.pop()
            used[i] = False


def permute(nums):
    used = [False] * len(nums)
    path = []
    out = []
    backtrack(nums, used, path, out)
    return out


def main():
    nums = [b + 1 for b in range(N)]

    acc = 0
    for k in range(TOTAL):
        nums[k % N] = 1 + (k % 9)
        perms = permute(nums)

        sig = 0
        for perm in perms:
            s = 0
            for i in range(len(perm)):
                s += perm[i] * (i + 1)
            sig = (sig * 31 + s) % MODULUS
        sig = (sig + len(perms)) % MODULUS
        acc = (acc * 131 + sig) % MODULUS

    print(acc)


if __name__ == "__main__":
    main()

File: leetcode/1-100/46-permutations/bench/permutations.rs
// LeetCode #46 bench mirror — Rust, the used-array mutable-path backtracker (★).
//
// Mirrors bench/permutations.kara exactly: a DFS that picks any still-unused element (tracked
// by a `used` bool array) alongside a mutable path, snapshotting the path into a Vec<Vec<i64>>
// at each leaf. Workload: TOTAL permutations of a fixed n=7 array with one slot punched per
// iteration (`nums[k%n] = 1 + k%9`), folding a position-weighted per-perm signature plus the
// count into a rolling checksum. Prints the same sink as every other mirror.

fn backtrack(
    nums: &[i64],
    used: &mut [bool],
    path: &mut Vec<i64>,
    out: &mut Vec<Vec<i64>>,
) {
    let n = nums.len();
    if path.len() == n {
        out.push(path.clone());
        return;
    }
    for i in 0..n {
        if !used[i] {
            used[i] = true;
            path.push(nums[i]);
            backtrack(nums, used, path, out);
            path.pop();
            used[i] = false;
        }
    }
}

fn permute(nums: &[i64]) -> Vec<Vec<i64>> {
    let n = nums.len();
    let mut used = vec![false; n];
    let mut path: Vec<i64> = Vec::new();
    let mut out: Vec<Vec<i64>> = Vec::new();
    backtrack(nums, &mut used, &mut path, &mut out);
    out
}

fn main() {
    let total: i64 = 300;
    let modulus: i64 = 1000000007;
    let n: i64 = 7;
    let mut nums: Vec<i64> = (0..n).map(|b| b + 1).collect();

    let mut acc: i64 = 0;
    for k in 0..total {
        nums[(k % n) as usize] = 1 + (k % 9);
        let perms = permute(&nums);

        let mut sig: i64 = 0;
        for perm in &perms {
            let mut s: i64 = 0;
            let mut i: i64 = 0;
            let pl = perm.len() as i64;
            while i < pl {
                s += perm[i as usize] * (i + 1);
                i += 1;
            }
            sig = (sig * 31 + s) % modulus;
        }
        sig = (sig + perms.len() as i64) % modulus;
        acc = (acc * 131 + sig) % modulus;
    }

    println!("{}", acc);
}
